SimPEG/regularization/coupling.py

```diff
@@ -0,0 +1,744 @@
+            term1 = temp1.dot(temp2)
+
+            temp1 = Dx_m1*Dx_m2 + Dy_m1*Dy_m2 + Dz_m1*Dz_m2
+            term2 = (np.linalg.norm(temp1))**2
+            result = term1 - term2
+
+            if normalized:
+                norms1, norms2, norms = self.gradient_applitude_inv(m1, m2, fltr=False)
+                temp1 = (Dx_m1*norms1)**2 + (Dy_m1*norms1)**2 + (Dz_m1*norms1)**2
+                temp2 = (Dx_m2*norms2)**2 + (Dy_m2*norms2)**2 + (Dz_m2*norms2)**2
+                term1 = temp1.dot(temp2)
+
+                temp1 = Dx_m1*Dx_m2*norms + Dy_m1*Dy_m2*norms + Dz_m1*Dz_m2*norms
+                term2 = (np.linalg.norm(temp1))**2
+                result = term1 - term2
+
+        return 0.5*result
+
+    def deriv(self, model):
+        '''
+        Computes the Jacobian of the cross-gradient.
+
+        :param list of numpy.ndarray ind_models: [model1, model2,...]
+
+        :rtype: numpy.ndarray
+        :return: result: gradient of the cross-gradient with respect to model1, model2
+
+        .. math::
+
+            See tutorial or documentation.
+
+        '''
+        m1 = self.map1*model
+        m2 = self.map2*model
+        m1, m2 = self.models([m1,m2])
+
+        if self.regmesh.mesh.dim == 2:
+            Dx = self.regmesh.aveFx2CC.dot(self.regmesh.cellDiffx)
+            Dy = self.regmesh.aveFy2CC.dot(self.regmesh.cellDiffy)
+
+            # common terms for dc_dm1
+            Dx_m1, Dy_m1 = Dx.dot(m1), Dy.dot(m1)
+            Dx_m2, Dy_m2 = Dx.dot(m2), Dy.dot(m2)
+            v1 = Dx_m2**2 + Dy_m2**2
+            v2 = Dx_m1**2 + Dy_m1**2
+            w = Dx_m1*Dx_m2 + Dy_m1*Dy_m2
+
+            dc_dm1 = (Dx.T.dot(Dx_m1*v1) + Dy.T.dot(Dy_m1*v1) -
+                      Dx.T.dot(Dx_m2*w) - Dy.T.dot(Dy_m2*w))
+            dc_dm2 = (Dx.T.dot(Dx_m2*v2) + Dy.T.dot(Dy_m2*v2) -
+                     Dx.T.dot(Dx_m1*w) - Dy.T.dot(Dy_m1*w))
+            result = np.concatenate((dc_dm1,dc_dm2))
+        elif self.regmesh.mesh.dim == 3:
+            Dx = self.regmesh.aveFx2CC.dot(self.regmesh.cellDiffx)
+            Dy = self.regmesh.aveFy2CC.dot(self.regmesh.cellDiffy)
+            Dz = self.regmesh.aveFz2CC.dot(self.regmesh.cellDiffz)
+
+            # common terms for dc_dm1
+            Dx_m1, Dy_m1, Dz_m1 = Dx.dot(m1), Dy.dot(m1), Dz.dot(m1)
+            Dx_m2, Dy_m2, Dz_m2 = Dx.dot(m2), Dy.dot(m2), Dz.dot(m2)
+            v1 = Dx_m2**2 + Dy_m2**2 + Dz_m2**2
+            v2 = Dx_m1**2 + Dy_m1**2 + Dz_m1**2
+            w = Dx_m1*Dx_m2 + Dy_m1*Dy_m2 + Dz_m1*Dz_m2
+
+            dc_dm1 = (Dx.T.dot(Dx_m1*v1) + Dy.T.dot(Dy_m1*v1) + Dz.T.dot(Dz_m1*v1) -
+                      Dx.T.dot(Dx_m2*w) - Dy.T.dot(Dy_m2*w) - Dz.T.dot(Dz_m2*w))
+            dc_dm2 = (Dx.T.dot(Dx_m2*v2) + Dy.T.dot(Dy_m2*v2) + Dz.T.dot(Dz_m2*v2) -
+                      Dx.T.dot(Dx_m1*w) - Dy.T.dot(Dy_m1*w) - Dz.T.dot(Dz_m1*w))
+            result = np.concatenate((dc_dm1,dc_dm2))
+
+        return result
+
+    def hessian_offdiag(self, D, grad1, grad2):
```

Comment:
maybe add a "v" kwarg that would avoid fully constructing the off-diagonal terms and just multiply by a vector.